compiler/daml-lf-tools/tests/DA/Daml/LF/Simplifier/Tests.hs

```diff
@@ -0,0 +1,118 @@
+import Test.Tasty
+import Test.Tasty.HUnit
+import qualified Data.NameMap as NM
+import qualified Data.Text as T
+
+import DA.Daml.LF.Ast.Base
+import DA.Daml.LF.Ast.Util
+import DA.Daml.LF.Ast.Version (versionDev)
+import DA.Daml.LF.Ast.World (initWorld)
+import DA.Daml.LF.Simplifier (simplifyModule)
+
+
+main :: IO ()
+main = defaultMain $ testGroup "DA.Daml.LF.Simplifier"
+    [ constantLiftingTests
+    ]
+
+constantLiftingTests :: TestTree
+constantLiftingTests = testGroup "Constant Lifting"
+    [ mkTestCase "empty module" [] []
+    , mkTestCase "closed value"
+        [ dval "foo" TInt64 (EBuiltin (BEInt64 10)) ]
+        [ dval "foo" TInt64 (EBuiltin (BEInt64 10)) ]
+    , mkTestCase "nested int"
+        [ dval "foo" (TInt64 :-> TInt64)
+            (ETmLam (ExprVarName "x", TInt64) (EBuiltin (BEInt64 10))) ]
+        [ dval "foo" (TInt64 :-> TInt64)
+            (ETmLam (ExprVarName "x", TInt64) (EBuiltin (BEInt64 10))) ]
+    , mkTestCase "nested arithmetic"
+        [ dval "foo" (TInt64 :-> TInt64)
+            (ETmLam (ExprVarName "x", TInt64)
+                (EBuiltin BEAddInt64
+                    `ETmApp` EBuiltin (BEInt64 10)
+                    `ETmApp` EBuiltin (BEInt64 10)))
+        ]
+        [ dval "$sc_foo_1" TInt64
+            (EBuiltin BEAddInt64
+                `ETmApp` EBuiltin (BEInt64 10)
+                `ETmApp` EBuiltin (BEInt64 10))
+        , dval "foo" (TInt64 :-> TInt64)
+            (ETmLam (ExprVarName "x", TInt64) (exprVal "$sc_foo_1"))
+        ]
+    , mkTestCase "\\xy.y" -- test that we aren't breaking up λxy.y into two lambdas.
+        [ dval "foo" (TInt64 :-> TInt64 :-> TInt64)
+            (ETmLam (ExprVarName "x", TInt64)
+                (ETmLam (ExprVarName "y", TInt64)
+                    (EVar (ExprVarName "y"))))
+        ]
+        [ dval "foo" (TInt64 :-> TInt64 :-> TInt64)
+            (ETmLam (ExprVarName "x", TInt64)
+                (ETmLam (ExprVarName "y", TInt64)
+                    (EVar (ExprVarName "y"))))
+        ]
+    , mkTestCase "\\z.(\\xy.y)z" -- test that we're lifting closed lambda subexpressions
+        [ dval "foo" (TInt64 :-> TInt64 :-> TInt64)
+            (ETmLam (ExprVarName "z", TInt64)
+                (ETmApp
+                    (ETmLam (ExprVarName "x", TInt64)
+                        (ETmLam (ExprVarName "y", TInt64)
+                            (EVar (ExprVarName "y"))))
+                    (EVar (ExprVarName "z"))))
+        ]
+        [ dval "$sc_foo_1" (TInt64 :-> TInt64 :-> TInt64)
+            (ETmLam (ExprVarName "x", TInt64)
+                (ETmLam (ExprVarName "y", TInt64)
+                    (EVar (ExprVarName "y"))))
+        , dval "foo" (TInt64 :-> TInt64 :-> TInt64)
+            (ETmLam (ExprVarName "z", TInt64)
+                (ETmApp
+                    (exprVal "$sc_foo_1")
+                    (EVar (ExprVarName "z"))))
+            -- NOTE: this is a candidate for eta reduction, may be optimized in the future
+        ]
+    ]
```

Comment:
Maybe it's worth adding one more case that somehow resembles lifting a computed dictionary, like to one for `Show [Int]`.

Reply:
This is a good idea!

Reply:
(I'm going to add this in a separate PR!)